### PR TITLE
CRYPTO

### PR DESCRIPTION
**Pull request by antonbobkov**

### CODE REVIEW

**Comment by antonbobkov** [comment]
did the thing where my_crypto can have simple/complex behavior depending on ifdef of a macro

**Comment by antonbobkov** [comment]
also fixed not signing things

File: src/my_crypto.erl [hunk 1]
```diff
@@ -1,26 +1,112 @@
 -module(my_crypto).
 
+-include_lib("eunit/include/eunit.hrl").
```
Comment (pdlla) [inline src/my_crypto.erl]: assert.hrl?

File: src/my_crypto.erl [hunk 2]
```diff
@@ -1,26 +1,112 @@
...
+
 -export([hash/1, sign/2, verify/3, read_file_key/2]).
 
--spec hash(iodata()) -> binary().
```
Comment (pdlla) [inline src/my_crypto.erl]: my -specs 😱

File: src/my_crypto.erl [hunk 3]
```diff
@@ -1,26 +1,112 @@
...
+
+%% Debugging simplified functions
+
+%% cut hash down to three bytes
+hash_debug(Bin) -> 
+    Full = crypto:hash(sha256, Bin),
+    Part = binary:list_to_bin(binary:bin_to_list(Full, 0, 3)),
+
+    %% Copied this online :cat_scream:
+    HexString = lists:flatten([io_lib:format("~2.16.0B",[X]) || <<X:8>> <= Part ]),
```
Comment (pdlla) [inline src/my_crypto.erl]: 🙀

File: src/my_crypto.erl [hunk 4]
```diff
@@ -1,26 +1,112 @@
...
+    Key.
+
+%% Debugging simplified functions
+
+%% cut hash down to three bytes
+hash_debug(Bin) -> 
+    Full = crypto:hash(sha256, Bin),
+    Part = binary:list_to_bin(binary:bin_to_list(Full, 0, 3)),
+
+    %% Copied this online :cat_scream:
```
Comment (pdlla) [inline src/my_crypto.erl]: :scream_cat:

Comment (pdlla) [inline src/my_crypto.erl]: : scream_cat :

File: src/my_crypto.erl [hunk 5]
```diff
@@ -1,26 +1,112 @@
...
-    [{_Key, _}] = public_key:ssh_decode(Bin, public_key),
-    %% Key;
-    FileName.
+    [{Key, _}] = public_key:ssh_decode(Bin, public_key),
+    Key.
+
+%% Debugging simplified functions
+
+%% cut hash down to three bytes
+hash_debug(Bin) ->
```
Comment (pdlla) [inline src/my_crypto.erl]: why is the name of this function hash_debug?

File: src/my_crypto.erl [hunk 6]
```diff
@@ -1,26 +1,112 @@
...
+drop_ext(FileName) ->
+    [Out | _] = string:split(FileName, "."),
+    Out.
+
+sign_debug(Hash, AnyKey) -> Hash ++ " " ++ drop_ext(AnyKey).
+verify_debug(Hash, Signature, PublicKey) -> Signature == sign_debug(Hash, PublicKey).
+read_file_key_debug(private, FileName) -> FileName;
+read_file_key_debug(public, FileName) -> FileName.
+
+-ifdef(MY_CRYPTO_DEBUG).
```
Comment (pdlla) [inline src/my_crypto.erl]: 😱

File: src/my_crypto.erl [hunk 7]
```diff
@@ -1,26 +1,112 @@
...
+
+hash(Bin) -> hash_debug(Bin).
```
Comment (pdlla) [inline src/my_crypto.erl]: what's the point of using hash_debug?

File: src/my_crypto.erl [hunk 8]
```diff
@@ -1,26 +1,112 @@
...
+
+hash(Bin) -> hash_full(Bin).
+sign(Hash, PrivateKey) -> sign_full(Hash, PrivateKey).
+verify(Hash, Signature, PubKey) -> verify_full(Hash, Signature, PubKey).
+read_file_key(private, FileName) -> read_file_key_full(private, FileName);
+read_file_key(public, FileName) -> read_file_key_full(public, FileName).
+
+-endif.
+
+%% -spec hash(iodata()) -> binary().
```
Comment (pdlla) [inline src/my_crypto.erl]: :D

File: src/my_crypto.erl [hunk 9]
```diff
@@ -1,26 +1,112 @@
...
+%% -spec sign(binary(), public_key:private_key()) -> binary().
+%% -spec verify(binary(), binary(), public_key:public_key()) -> boolean().
+%% -spec read_file_key(private, string()) -> _.
+
+my_crypto_test() ->
```
Comment (pdlla) [inline src/my_crypto.erl]: move to my_crypto_test.erl?

File: test/pop_manager_test.erl [hunk 10]
```diff
@@ -81,13 +85,15 @@ basic_test() ->
...
+generate_chain(PC0, NewBlockCount, InitialVerifierNum, PrivateKey) ->
 
     MapAccFn = fun(PC, VerNum) ->
 		       Block = pop_chain:generate_new_block(VerNum, PC),
 		       AddTime = get_block_cd(timestamp, Block),
 
-		       {pop_chain:add_block_in_order(Block, AddTime, PC), Block}
+		       SignedBlock = pop_chain:apply_block_signature(my_crypto:sign(maps:get(this_id, Block), PrivateKey), Block),
+
+		       {pop_chain:add_block_in_order(SignedBlock, AddTime, PC), SignedBlock}
```
Comment (pdlla) [inline test/pop_manager_test.erl]: :o such block signing

File: src/my_crypto.erl [hunk 11]
```diff
@@ -1,26 +1,112 @@
 -module(my_crypto).
```
Comment (pdlla) [inline src/my_crypto.erl]: rename to potato_crypto?

File: src/my_crypto.erl (hunk 4 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: literally unplayable

File: src/my_crypto.erl (hunk 6 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: gonna change it to POTATO_CRYPTOGRAPHY_MODE

File: src/my_crypto.erl (hunk 7 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: its shorter. don't need 32 numbers clogging up the output

File: src/my_crypto.erl (hunk 8 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: specs too hard :scream_cat:

File: src/my_crypto.erl (hunk 9 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: didn't want to export too many functions: hash_debug etc

this way i only export useful stuff

File: test/pop_manager_test.erl (hunk 10 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: much security o:

File: src/my_crypto.erl (hunk 11 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: YA

File: src/my_crypto.erl (hunk 9 above)
Comment (pdlla) [reply src/my_crypto.erl]: such privacy

File: src/my_crypto.erl (hunk 7 above)
Comment (pdlla) [reply src/my_crypto.erl]: :o how bad could it be?

Comment (pdlla) [reply src/my_crypto.erl]: when potato game gets big, someone will file a ticket

"signing hash collision in potato cryptography mode"

**Comment by antonbobkov** [comment]
added some more testing code!! its pretty boring, feel free to ignore

File: test/pop_manager_test.erl [hunk 12]
```diff
@@ -115,37 +121,80 @@ extract_new_block_messages() ->
...
+    {PC_C, [C1, C2, C3, C4, C5]} = generate_chain(PC_A, 5, 1, PrivateKey),
+
+    %% We have constructed test chain:
+    %% 
+    %%                B1 - B2 - B3 - B4
+    %%              /
+    %% A0 - A1 - A2 
+    %%              \
+    %%                C1 - C2 - C3 - C4 - C5
+    %%
```
Comment (antonbobkov) [inline test/pop_manager_test.erl]: except for this part, this part is awesome

File: src/my_crypto.erl (hunk 7 above)
Comment (antonbobkov) [reply src/my_crypto.erl]: just for debugging

on error I usually get a giant red wall of text dumped on my screen

this makes it a bit more managable

File: test/pop_manager_test.erl (hunk 12 above)
Comment (pdlla) [reply test/pop_manager_test.erl]: good except needs more fish and salad emojis  🥗 🐟

File: test/pop_manager_test.erl [hunk 13]
```diff
@@ -115,37 +121,80 @@ extract_new_block_messages() ->
...
+	{net, send_full_blocks, {old, [Block]}} -> 
+	    [Block | extract_send_full_blocks()];
+	_Any ->
+	    fail
+    after 0 -> 
+	    []
+    end.
+    
+
 no_more_messages() ->
```
Comment (pdlla) [inline test/pop_manager_test.erl]: cool function

File: test/pop_manager_test.erl [hunk 14]
```diff
@@ -115,37 +121,80 @@ extract_new_block_messages() ->
...
+	   player_id => PublicKey,
+	   consensus_data => #{
+			       signature => undefined,
+			       chain_id => ChainId
+			      }
+	  },
+    Hash = my_crypto:hash(my_serializer:serialize_object(T0)),
+    Signature = my_crypto:sign(Hash, PrivateKey),
+    CD = maps:get(consensus_data, T0),
+    T1 = T0#{
```
Comment (pdlla) [inline test/pop_manager_test.erl]: should there be a helper function to hash + sign a block?

File: test/pop_manager_test.erl [hunk 15]
```diff
@@ -115,37 +121,80 @@ extract_new_block_messages() ->
...
+    %%              /
+    %% A0 - A1 - A2 
+    %%              \
+    %%                C1 - C2 - C3 - C4 - C5
+    %% 
 
-    {PC_C, [C1, C2, C3, C4, C5]} = generate_chain(PC_A, 5, 1),
 
     PM_B = PM#pop_manager{pop_chain = PC_B},
+    PM_C = PM#pop_manager{pop_chain = PC_C},
```
Comment (pdlla) [inline test/pop_manager_test.erl]: So you can have pop_managers on different chains? 
When does pop manager pick a fork?

File: test/pop_manager_test.erl [hunk 16]
```diff
@@ -193,5 +243,52 @@ fancy_test() ->
 
     ?assert(no_more_messages()),
 
+    %% Test request_block_hash_range
+
+    ReqHashRangeTestFn = 
+	fun(RequestList, State, ResultList) ->
+		StateM = pop_manager:on_net_message(self(), Time, request_block_hash_range, get_hashes_tuple(RequestList), State),
+
+		?assertEqual(StateM, State),
```
Comment (pdlla) [inline test/pop_manager_test.erl]: Is this is to ensure request_block_hash_range doesn't change pop_manager state?

File: test/pop_manager_test.erl [hunk 17]
```diff
@@ -193,5 +243,52 @@ fancy_test() ->
...
+
+		?assertEqual({net, send_block_hashes, get_hashes(ResultList)}, extract_message()),
```
Comment (pdlla) [inline test/pop_manager_test.erl]: :o, cool!

File: test/pop_manager_test.erl (hunk 14 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: eh, probably at some point

File: test/pop_manager_test.erl (hunk 12 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: OMG EACH HASH NEEDS TO BE AN EMOJI

gonna rewrite my hash_debug function

File: test/pop_manager_test.erl (hunk 16 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: ya, pretty clear from the code that it doesn't but just a quick sanity check

File: test/pop_manager_test.erl (hunk 15 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: line 210 - I construct PM_BC that has the fork

File: test/pop_manager_test.erl (hunk 13 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: zero timer hype

File: test/pop_manager_test.erl (hunk 12 above)
Comment (antonbobkov) [reply test/pop_manager_test.erl]: https://gist.github.com/rxaviers/7360908

:money_with_wings: :crystal_ball: :waning_crescent_moon: :water_buffalo: :stuck_out_tongue_closed_eyes:

Comment (antonbobkov) [reply test/pop_manager_test.erl]: https://api.github.com/emojis

Comment (pdlla) [reply test/pop_manager_test.erl]: watch out for emoji collisions :o!

Comment (pdlla) [reply test/pop_manager_test.erl]: https://www.youtube.com/watch?v=u011XHmYq1Q

Comment (antonbobkov) [reply test/pop_manager_test.erl]: SHA256 hash is equivalent to approx 11 emojis

**Pull request merged**